### ISSUE TITLE
Feature request: Autolocking database when app loses focus

### ISSUE THREAD

**Issue by ruvilonix** [Kunzisoft/KeePassDX#1502]
The Standard Notes app has a setting for the app to lock the moment it is not the active app. It would be nice to have this for KeePassDX as well.

Also, the idle timeout setting should not be changeable if not logged in to the database. The way it currently is, someone with access to the phone could increase or turn off the idle timeout without the database owner knowing. Next time the owner unlocks the database, they may assume it will autolock after a few minutes. It doesn't and the other person has full access if they are able to access the phone again..

**Comment by J-Jamet** [comment]
> The Standard Notes app has a setting for the app to lock the moment it is not the active app. It would be nice to have this for KeePassDX as well.

It's worth looking into, but the big drawback is that it would make the Magikeyboard and Autofill not very usable, and a password manager that can't fill in passwords is annoying.

> Also, the idle timeout setting should not be changeable if not logged in to the database.

For this, the application simply relies on the fact that the device can be secured. Do not hesitate to change the scheme of your device or to put another secure unlocking mode.

**Comment by ruvilonix** [comment]
> It's worth looking into, but the big drawback is that it would make the Magikeyboard and Autofill not very usable, and a password manager that can't fill in passwords is annoying.

That's true, but I think it should be an option for people who are more concerned with security, or aren't that worried about the autofill. For example, I mostly memorize all my passwords, but there are times that I forget, so I open up the password manager to check.

> For this, the application simply relies on the fact that the device can be secured. Do not hesitate to change the scheme of your device or to put another secure unlocking mode.

What's the point of the idle timeout then if not for security? People might share their phone with others, even if just for a few minutes to show them pictures.

**Comment by J-Jamet** [comment]
If the need is to close the application by pressing the back button, then it is already implemented in Settings - App settings - press 'Back' to lock.

> People might share their phone with others, even if just for a few minutes to show them pictures.

I have never understood this practice. :camera: 
But if you want to add a protection at the level of the app opening, then I recommend you to use a launcher which allows to put an applock on any application.

**Comment by ruvilonix** [comment]
> If the need is to close the application by pressing the back button, then it is already implemented in Settings - App settings - press 'Back' to lock.

But that's also able to be disabled without having the database unlocked. If that setting could be moved to only be editable with the password, I think it would be a good alternative to autolock.

**Comment by J-Jamet** [comment]
> But that's also able to be disabled without having the database unlocked. If that setting could be moved to only be editable with the password, I think it would be a good alternative to autolock.

=> Use a launcher with an app lock, I put it just before.
The settings of a database are related to the database, not to the app.

**Comment by ruvilonix** [comment]
> => Use a launcher with an app lock, I put it just before. 

I'll look into the app lock. So what is the point of the idle timeout and lock on back button setting for? The only thing they would be useful for is preventing other users of the phone from seeing passwords. But the settings are changeable by those other users, they seem to do pretty much nothing. 

> The settings of a database are related to the database, not to the app.

Couldn't that setting be added to the database?

**Comment by J-Jamet** [comment]
> So what is the point of the idle timeout and lock on back button setting for?

To lock your database.

> But if the settings are changeable by those other users, they seem to do pretty much nothing.

You are aware that if you give an unlocked phone to someone, he can send messages for you, change all the settings of the device, install and delete applications, change and delete files.
If your database is locked, changing the app settings won't magically allow them to open it, but the attacker may have found tons of other methods to retrieve your master key when you re-enter the credential. So lock down and don't give away your unlocked device.

**Comment by ruvilonix** [comment]
It seems like if the point of the idle timeout setting is in any way a method of increasing security, it would be worth it to move editing it to behind the password to increase that security further. It doesn't seem like it would be very difficult to do this. Add a line into the database file: idletimeout = 30s. When Keepass opens the database, it checks that line and enables the timeout accordingly.

**Comment by J-Jamet** [comment]
We need to set up this new mechanism and find a way to inform the user that he can save these settings in the database and that the global settings can be overwritten according to those in the database. This is possible but not easy to implement, especially for the UX.
